proposals/0000-quick-look-impredicativity.md

```diff
@@ -0,0 +1,181 @@
+
+A few versions ago, the answer would have been **no**. The reason was that `($)` was the only function able to operate with both lifted and unlifted kinds. But nowadays, this fact is reflected in its type:
+
+```haskell
+($) ::
+  forall (r :: GHC.Types.RuntimeRep) a (b :: TYPE r).
+  (a -> b) -> a -> b
+```
+
+So the initial answer would be **yes**. This should be taken with a grain of salt, though, since the specification in the draft paper does not consider polymorphic type representations. However, for the usual case of `b` having kind `*`, the rule may be dropped. This also means that other combinators such as `(&)` or `(.)` no longer are second-class with respect to impredicativity.
+
+## Costs and Drawbacks
+
+As discussed below, there is already a branch of GHC in which these changes have been implemented. Furthermore, we expect maintenance costs to be low, since "quick look" is a separate phase from the rest of type checking; there is no intrincate relation between the two phases in the code.
+
+`ImpredicativeTypes` has always been an advanced feature of the language. However, once you arrive at it, programmers often ask themselves: "what is stopping the compiler from accepting this?". This proposal gives an answer to that question, with a succint explanation: "impredicative is never guessed, it must be obvious from the simple parts of each application".
+
+The main drawback is the loss of contravariance that this proposal brings with it, unless an extension is enabled. It is arguable whether contravariance of function types is something a beginner (or even intermediate) Haskeller expects, since up to that point all types are compared by equality. However, it should be possible during error reporting to scan whether a type error talks about two types which could be related if contravariance was taking into account, and report a message on the lines of:
+
+```text
+Hint: try to eta-expand the second argument.
+```
+
+## Alternatives
+
+The main alternative is to keep the *status quo*: `ImpredicativeTypes` is there, but never officially supported. However, that may bring surprising behavior, which may change unexpectedly with a new GHC release.
+
+Another possibility is to drop impredicative type inference altogether, and expect the programmers to use `TypeApplications` anytime it is required. The problem is that impredicative instantiation is quite contagious, so one might need to write things like:
+
+```haskell
+(++) @(forall a. a -> a) ids ([] @(forall a. a -> a))
+```
+
+Finally, we could also enable `ContravariantFunctions` whenever `RankNTypes` is enabled but `ImpredicativeTypes` is not. The main disadvantage is that code that compiles under the former flag may stop compiling if you add the latter.
+
+## Unresolved Questions
+
+* What should be the error messages in those cases in which "quick look" was unsuccessful and did not infer enough information for impredicative instantiation?
```

Comment:
I'm afraid that I don't understand this question.

Reply:
No, but I thought that would be useful during the discussion.

Reply:
Rephrased it. But the main idea is: what should we do with error messages related to impredicative instantiation? can we do better at diagnosing problems and suggesting fixes?